[serve] Rolling updates for redeployments

edoakes opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Performs a rolling update when re-deploying a backend. Currently, the batch size of the update is hardcoded to floor(20%) of the target number of replicas. We can make this configurable in the future.

This follows the same protocol as [k8s](https://kubernetes.io/docs/tutorials/kubernetes-basics/update/update-intro/):

- Old replicas are turned down before new ones are started.
- At most batch_size old replicas will be stopped and new replicas will be started at once.

## Related issue number

Closes https://github.com/ray-project/ray/issues/14805

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

simon-mo commented:
It's going to take sometime to review this, ETA tuesday noon.

edoakes commented on python/ray/serve/backend_state.py:
```diff
@@ -596,22 +621,35 @@ def delete_backend(self, backend_tag: BackendTag,
                 ReplicaState.SHOULD_START, ReplicaState.STARTING,
                 ReplicaState.RUNNING
             ])
+        new_running_replicas = self._replicas[backend_tag].count(
+            version=target_version, states=[ReplicaState.RUNNING])
+
+        pending_replicas = (
+            target_replicas - new_running_replicas - old_running_replicas)
+        rollout_size = max(int(0.2 * target_replicas), 1)
+
+        max_to_stop = max(rollout_size - pending_replicas, 0)
+        replicas_to_stop = self._replicas[backend_tag].pop(
+            exclude_version=target_version,
+            states=[
+                ReplicaState.SHOULD_START, ReplicaState.STARTING,
+                ReplicaState.RUNNING
+            ],
+            max_replicas=max_to_stop)
```
@simon-mo this the core logic to review FYI

simon-mo commented on python/ray/serve/api.py:
```diff
@@ -447,7 +452,8 @@ def deploy(self,
                config: Optional[Union[BackendConfig, Dict[str, Any]]] = None,
-               version: Optional[str] = None) -> None:
+               version: Optional[str] = None,
+               _blocking: Optional[bool] = True) -> None:
```
change the return type? it's not Optional[GoalRef]?

simon-mo commented on python/ray/serve/tests/test_backend_state.py:
```diff
@@ -134,16 +134,40 @@ def version(self):
+        states=[ReplicaState.STARTING, ReplicaState.STOPPING])
+    assert c.count(states=[ReplicaState.STARTING]) == 2
+    assert c.count(states=[ReplicaState.STOPPING]) == 1
+    assert not c.count(states=[ReplicaState.SHOULD_START])
+    assert not c.count(
+        states=[ReplicaState.SHOULD_START, ReplicaState.SHOULD_STOP])
```
`== 0` to be more explicit

simon-mo commented on python/ray/serve/tests/test_backend_state.py:
```diff
@@ -616,32 +636,298 @@ def test_deploy_new_config_new_version(mock_backend_state):
+    # Now deploy a new version. Two old replicas should be stopped.
+    goal_2 = backend_state.deploy_backend(
+        tag, b_config_1, r_config_1, version="2")
+    backend_state.update()
+    assert backend_state._replicas[tag].count() == 10
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.RUNNING]) == 8
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.STOPPING]) == 2
+    assert not goal_manager.check_complete(goal_2)
+
+    backend_state.update()
+    assert backend_state._replicas[tag].count() == 10
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.RUNNING]) == 8
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.STOPPING]) == 2
+    assert not goal_manager.check_complete(goal_2)
```
These two blocks are repeated? if intentional put them in for _ in range(2)

edoakes replied:
oops, accidental

simon-mo commented on python/ray/serve/tests/test_backend_state.py:
```diff
@@ -616,32 +636,298 @@ def test_deploy_new_config_new_version(mock_backend_state):
+        states=[ReplicaState.STOPPING])[0]._actor.set_done_stopping()
+
+    backend_state.update()
+    assert backend_state._replicas[tag].count() == 9
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.RUNNING]) == 8
+    assert backend_state._replicas[tag].count(
+        version="1", states=[ReplicaState.STOPPING]) == 1
```
i think you can abstract these chunk a bit more to help us understand the test.
```python
check_state(
    {
        "total": 9,
        ("1", ReplicaState.RUNNING): 8, 
        ...
    }
)
```

edoakes replied:
good idea, done!

simon-mo commented on python/ray/serve/backend_state.py:
```diff
@@ -596,22 +621,35 @@ def delete_backend(self, backend_tag: BackendTag,
-        if version is None:
+        if target_version is None:
             return
 
```
do we have any check that target_replica cannot be 0?

edoakes replied:
target_replicas can be 0, this happens in the case when we're deleting a backend. in that case, we just want to tear everything down anyways so it doesn't matter how many we stop here.

simon-mo commented on python/ray/serve/backend_state.py:
```diff
@@ -596,22 +621,35 @@ def delete_backend(self, backend_tag: BackendTag,
+        new_running_replicas = self._replicas[backend_tag].count(
+            version=target_version, states=[ReplicaState.RUNNING])
+
+        pending_replicas = (
```
we can assert that this is equal to .count(version=target_version, states=[SHOULD_START,STARTING])?

architkulkarni replied:
It might be more manageable if we separate the rolling upgrade logic from the scaling logic instead of doing them both in _scale_backend_replicas

edoakes replied:
So the problem here is that once replicas are completely stopped, they are removed from the state data structure, which means you can't use the equality check that @architkulkarni suggested (you could make it a <= though). That's why I'm "reverse engineering" the pending replicas instead of directly counting the number stopping and number starting.

Good point about the negative pending_replicas @architkulkarni. I agree that it would be better to separate this logic out, let me see if there's a clean way to do that. It'd also help with unit testing for both methods.

architkulkarni commented on python/ray/serve/backend_state.py:
```diff
@@ -596,22 +621,35 @@ def delete_backend(self, backend_tag: BackendTag,
+        old_running_replicas = self._replicas[backend_tag].count(
+            exclude_version=target_version,
             states=[
                 ReplicaState.SHOULD_START, ReplicaState.STARTING,
```
We want to ensure there are never more than rollout_size replicas not running right?  So shouldn't pending_replicas be called not_running_replicas, and shouldn't old_version SHOULD_START and STARTING contribute to not_running_replicas?

edoakes replied:
Yep, good point.

architkulkarni replied:
I see you added a comment here.  It makes sense to include SHOULD_START and STARTING at the bottom where we actually pop replicas_to_stop, which is how it currently is.  I think that's the part your comment refers to.

But I still don't think it makes sense to include SHOULD_START and STARTING here.  We have `pending_replicas = target_replicas - new_running_replicas - old_running_replicas`.  The meaning of `pending_replicas` seems to be the number of non-running replicas of any version, which is why `pending_replicas` contributes to the `rollout_size` cap.

architkulkarni replied:
Ah never mind, I understand now.

architkulkarni commented on python/ray/serve/backend_state.py:
```diff
@@ -596,22 +621,35 @@ def delete_backend(self, backend_tag: BackendTag,
+
+        pending_replicas = (
+            target_replicas - new_running_replicas - old_running_replicas)
+        rollout_size = max(int(0.2 * target_replicas), 1)
```
Would be good to add a comment saying that rollout_size is the max number of replicas we ever want to be not running any version

architkulkarni reviewed (commented):
I looked at the core logic and I had a couple of concerns, but I might be misunderstanding or misremembering something.  I think it's close though!  Overall more comments would be helpful.

edoakes commented on python/ray/serve/backend_state.py:
```diff
@@ -596,22 +621,35 @@ def delete_backend(self, backend_tag: BackendTag,
     def _stop_wrong_version_replicas(
-            self, backend_tag: BackendTag, version: str,
-            graceful_shutdown_timeout_s: float) -> int:
+            self, backend_tag: BackendTag, target_replicas: int,
```
Yeah this should definitely be documented better. Let me clean it up a bit.

edoakes commented:
@simon-mo @architkulkarni I updated the tests to be easier to read and added two new test cases. I also addressed the issue @architkulkarni pointed out about target_replicas being below the current pending_replicas & added some more comments explaining the logic.

Please take another look and let me know what's still confusing / if you see any issues in the logic.

simon-mo commented on python/ray/serve/backend_state.py:
```diff
@@ -596,40 +621,70 @@ def delete_backend(self, backend_tag: BackendTag,
+        # or rollout_size new replicas starting.
+        rollout_size = max(int(0.2 * target_replicas), 1)
+        max_to_stop = max(rollout_size - pending_replicas, 0)
+
```
I think we still need to handle the edge of target_replicas = 0 here:
```
target_replicas = 0
old_running_replicas == old_stopping_replicas == 0
new_running_replicas = 0
pending_replicas = 0 - 0 - 0 = 0

rollout_size = max(int(0.2*0), 1) = 1
max_to_stop = max(1 - 0, 0) = 1
```
but max_to_stop should equal 0.

edoakes replied:
The behavior in your example should be fine because there won't be any replicas to stop anyways (it's a `max`, not an exact number to stop), but yeah we should explicitly handle this. I will short-circuit if target_replicas is 0.

simon-mo commented on python/ray/serve/backend_state.py:
```diff
@@ -596,40 +621,70 @@ def delete_backend(self, backend_tag: BackendTag,
+                ReplicaState.RUNNING
+            ],
+            max_replicas=max_to_stop)
 
```
in this version, seems like we are not counting number of new version replicas that are starting? namely if the following is > 0, we are breaking the invariant the max number of replicas in transition.
```
replicas.count(
            version=target_version, states=[ReplicaState.STARTING, ReplicaState.SHOULD_START])
```

architkulkarni replied:
I think this part might be okay--the expression you've written is contributes to pending_replicas.

Concretely, if `target_replicas = current_replicas =` 10 replicas, and it consists of 8 old RUNNING and 2 new STARTING, then we have 

`pending_replicas = 10 - 8 - 0 = 2`

so `max_to_stop` is `2-2 = 0` and nothing gets stopped.  (Until the new_version replicas move from STARTING to RUNNING, as desired.)

simon-mo commented on python/ray/serve/tests/test_backend_state.py:
```diff
@@ -596,52 +652,642 @@ def test_deploy_new_config_new_version(mock_backend_state):
+        states=[ReplicaState.STARTING])[0]._actor.set_ready()
+    backend_state.update()
+
+    backend_state.update()
+    check_counts(backend_state, total=10)
+    check_counts(
+        backend_state,
+        version="1",
+        total=9,
+        by_state=[(ReplicaState.RUNNING, 7), (ReplicaState.STOPPING, 2)])
+    check_counts(
+        backend_state,
+        version="2",
+        total=1,
+        by_state=[(ReplicaState.RUNNING, 1)])
+
+    backend_state.update()
+    check_counts(backend_state, total=10)
+    check_counts(
+        backend_state,
+        version="1",
+        total=9,
+        by_state=[(ReplicaState.RUNNING, 7), (ReplicaState.STOPPING, 2)])
+    check_counts(
+        backend_state,
+        version="2",
+        total=1,
+        by_state=[(ReplicaState.RUNNING, 1)])
```
these two chunks are duplicated

simon-mo replied:
move them to for loop?

edoakes replied:
just removed the redundancy

architkulkarni reviewed (approved):
Logic and new tests look good!  The comments are helpful.  I think the core logic would still be hard to understand for someone new to the code but I can't really think of a way to make it much clearer.  One idea might be to move some of the algebra into the algorithm itself--specifically, actually pop all the replicas we can stop for "free" (namely, old SHOULD START and old STARTING) at the beginning, so we never have to account for them in the rest of the function.  In any case, that can be done in a future PR (if at all), and the tests you wrote will make it much easier.

edoakes commented:
@architkulkarni I totally agree, I'm not very happy with the readability/understandability, but I couldn't find a better way to express it. I tried splitting out the "rolling update" and "regular scaling" logic into separate codepaths, but it ended up being both hard to make correct and added some redundancy. Hopefully we can find a way to clean this up in the near future.
